winch/codegen/src/frame/mod.rs

```diff
@@ -101,49 +101,87 @@ impl Frame {
                     )),
-                    align_to(vmctx_offset, <A as ABI>::stack_align().into()),
+                    defined_locals_end,
                 ),
-                ABIOperand::Reg { ty, .. } => {
-                    let offs = align_to(vmctx_offset, ptr_size) + ptr_size;
+                // If the resuls operand is a register, we give this register
+                // the same treatment as all the other argument registers and
+                // spill it, therefore, we need to increase the locals size by
+                // one slot.
+                ABIOperand::Reg { ty, size, .. } => {
+                    let offs = align_to(defined_locals_end, *size) + *size;
                     (
                         Some(LocalSlot::new(*ty, offs)),
                         align_to(offs, <A as ABI>::stack_align().into()),
                     )
                 }
             }
         } else {
-            (
-                None,
-                align_to(vmctx_offset, <A as ABI>::stack_align().into()),
-            )
+            (None, defined_locals_end)
         };
 
+        let vmctx_slot = *locals.get(0).expect("LocalSlot for VMContext");
         Ok(Self {
             locals,
             locals_size,
-            vmctx_slot: LocalSlot::i64(vmctx_offset),
+            vmctx_slot,
             defined_locals_range: DefinedLocalsRange(
                 defined_locals_start..(defined_locals_start + defined_locals.stack_size),
             ),
             results_base_slot,
         })
     }
 
-    /// Get a local slot.
-    pub fn get_local(&self, index: u32) -> Option<&LocalSlot> {
-        self.locals.get(index as usize)
+    // Winch's ABI uses two extra parameters to store the callee and caller
+    // VMContext pointers.
+    // These arguments are spilled and treated as frame locals, but not
+    // WebAssembly locals.
+    const WASM_LOCALS_OFFSET: usize = 2;
+
+    /// Get the [LocalSlot] for a WebAssembly local.
+    /// This method assumes that the index is bound to u32::MAX, representing
+    /// the index space for WebAssembly locals.
+    ///
+    /// # Panics
+    /// This method panics if the index is not associated to a valid WebAssembly
+    /// local.
+    pub fn get_local(&self, index: u32) -> &LocalSlot {
+        let local_index = Self::WASM_LOCALS_OFFSET + index as usize;
+        self.locals
+            .get(local_index)
+            .unwrap_or_else(|| panic!(" Expected WebAssembly local at slot: {index}"))
+    }
+
+    /// Get the [LocalSlot] for a frame local.
+    /// This method doesn't make any asumptions about the local index passed in,
+    /// and simply delegates the [LocalSlot] retrieval to the underlying locals
+    /// vector.
+    ///
+    /// # Panics
+    /// This method panics if the index is not associated to a valid WebAssembly
+    /// local.
+    pub fn get_local_unchecked(&self, index: usize) -> &LocalSlot {
```

Comment:
Rather than `unchecked`, which to me implies lack of a bounds check or similar, can we name this in a way that indicates the index-space is shifted relative to above? E.g. perhaps the above `get_local` becomes `get_wasm_local` and `get_local_unchecked` becomes `get_raw_local`?

Alternately, avoid exposing `get_local_unchecked` at all and have two separate methods `get_callee_local()` / `get_caller_local()`?

Reply:
Thanks for the suggestion -- I ended up with `get_frame_local` and `get_wasm_local` to be able to distinguish one from another.